Add transpiler pass for complementary control pattern simplification using Boolean algebra

Mostafa-Atallah2020 opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR adds a new transpiler pass `ControlPatternSimplification` that optimizes multi-controlled gates with complementary control patterns using Boolean algebraic simplification.

**Note:** This is an initial skeleton/draft PR. Full implementation will follow in subsequent commits.

Fixes #15348

### Details and comments

#### Motivation

Multi-controlled gates with different control patterns can often be simplified when the patterns are complementary. For example, gates controlled on `'11'` and `'01'` both require q0=1, so they can be combined into a single gate controlled only on q0.

This pattern appears frequently in:
- Hamiltonian simulation
- Variational algorithms
- Quantum walk circuits

As demonstrated in #15348, the transpiler currently produces 31 gates for a case that could be optimized to just 5 gates (520% overhead).

#### Optimization Techniques

The pass will support four main optimization strategies:

1. **Complementary patterns**: `(q0 ∧ q1) ∨ (q0 ∧ ¬q1) = q0`
   - Example: `['11', '01']` → single control on q0
   - Reduces 2 multi-controlled gates to 1 single-controlled gate

2. **Subset patterns**: `(q0 ∧ q1 ∧ q2) ∨ (q0 ∧ q1 ∧ ¬q2) = (q0 ∧ q1)`
   - Example: `['111', '110']` → reduces control count
   - Removes redundant control qubits

3. **XOR pairs**: Patterns satisfying `q1 ⊕ q2 = 1`
   - Example: `['110', '101']` → CNOT-based optimization
   - Reduces 2 multi-controlled gates to 1 + 2 CNOTs

4. **Complete partitions**: All control states covered
   - Example: `['00','01','10','11']` → unconditional gate
   - Eliminates all control qubits

#### Example Usage

```python
from qiskit import QuantumCircuit
from qiskit.circuit.library import RXGate
from qiskit.transpiler.passes import ControlPatternSimplification

theta = np.pi / 4
qc = QuantumCircuit(3)
qc.append(RXGate(theta).control(2, ctrl_state='11'), [0, 1, 2])
qc.append(RXGate(theta).control(2, ctrl_state='01'), [0, 1, 2])

pass_ = ControlPatternSimplification()
optimized = pass_(qc)  # Result: Single CRX gate controlled by q0
```

#### Files Changed

- `qiskit/transpiler/passes/optimization/control_pattern_simplification.py` - Main pass implementation (skeleton)
- `qiskit/transpiler/passes/optimization/__init__.py` - Export pass
- `qiskit/transpiler/passes/__init__.py` - Documentation and import
- `test/python/transpiler/test_control_pattern_simplification.py` - Test skeleton

#### References

- Atallah et al., "Graph Matching Trotterization for Continuous Time Quantum Walk Circuit Simulation", Proceedings of IEEE QCE 2025
- Atallah et al., "Simulating Quantum Walk Hamiltonians without Pauli Decomposition", arXiv:2601.11418 (2026)
- Gonzalez et al., "Efficient sparse state preparation via quantum walks", npj Quantum Information (2025)
- Amy et al., "Fast synthesis of depth-optimal quantum circuits", IEEE TCAD 32.6 (2013)
- Shende & Markov, "On the CNOT-cost of TOFFOLI gates", arXiv:0803.2316 (2008)
- Barenco et al., "Elementary gates for quantum computation", Phys. Rev. A 52.5 (1995)

#### Next Steps

- [x] Implement core optimization logic
- [x] Add comprehensive test cases
- [x] Benchmark against common use cases
- [x] Add release notes using reno

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=15352) <br/>All committers have signed the CLA.

debasmita2102 commented on qiskit/transpiler/passes/optimization/control_pattern_simplification.py:
```diff
@@ -0,0 +1,126 @@
+from qiskit.utils import optionals as _optionals
+
+
+@_optionals.HAS_SYMPY.require_in_instance
```
I assume SymPy would only be useful if you are doing complex symbolic Boolean algebra simplification with variables. But your control patterns are concrete bit strings right? Not symbolic expressions. 
For example : **Complementary pattern detection:** `['11', '01'] →` requires string/bitwise operations, not SymPy, isn't it? Similarly a few are there.
This decorator forces SymPy to be installed to even import the class.  I guess, SymPy can make the pass harder to install and slower to import, can you recheck?

Mostafa-Atallah2020 replied:
Thanks @debasmita2102, You're absolutely right that the control patterns are concrete bit strings. I initially chose SymPy's `simplify_logic` because it provides a general-purpose Boolean algebra simplifier that automatically handles all the optimization cases.

However, I completely agree that adding SymPy as a dependency for concrete bit string operations is not ideal. I will work on refactoring the code to use bitwise operations instead, which should eliminate the SymPy dependency entirely while maintaining the same optimization capabilities. I'll update the PR once this is done.

patelvyom replied:
You can avoid using SymPy by leveraging the existing ESOP-based simplification logic in `qiskit/synthesis/boolean/boolean_expression_synth.py`. What you need is essentially a variant of `synth_bit_oracle_from_esop`, except that the operation applied to the target qubit is not an `X` gate but an arbitrary `Gate`, provided, of course, that the same gate is used for every control-bit pattern.

patelvyom replied:
As a simple example for ['11', '01']:
```
expr = BooleanExpression("x0 & x1 | ~x0 & x1")
esop = EsopGenerator(expr.truth_table).esop
print(esop)
['-1']
```
where `-` indicates no control on `x0`. So `esop` can be used to create a new circuit with appropriate control patterns.

Mostafa-Atallah2020 replied:
I tried to get rid of sympy but it didn't work out actually

Cryoris replied:
What was the issue here? If we don't have to rely on SymPy that would be great -- at the very least it's import needs to be made optional since we don't have a hard dependency on it.

Mostafa-Atallah2020 replied:
Hi @Cryoris the reason why i am using sympy is because this code was part of another project I worked on. I tried to remove sympy dependency as @debasmita2102 suggested but all the tests are ended up being failing. I just can't figure out how to implement this without sympy.

debasmita2102 commented on test/python/transpiler/test_control_pattern_simplification.py:
```diff
@@ -0,0 +1,93 @@
+
+        pass_ = ControlPatternSimplification()
+        # optimized = pass_(qc)
+
```
Should gates with different parameters like `RXGate(0.5).control(2, ctrl_state='11')` and `RXGate(0.7).control(2, ctrl_state='01')` merge ? Will it produce correct quantum circuits? I guess not. In that case we can add something like `test_different_parameters_no_merge() ` before implementing optimization logic.

alexanderivrii replied:
@debasmita2102, I agree, I don't immediately see whether the two gates above can be combined. Presumably we should check that the base gates are exactly the same, including parameters, and I agree that we should have a test for that.

This makes me think that we also don't have a way to combine controlled gates _with the same control state_ in the case that the base gates can be combined, e.g. in principle we could combine `RXGate(0.5).control(2, crtl_state='10')` and `RXGate(0.7).control(2, crtl_state='10')` into `RXGate(1.2).control(2, crtl_state='10')`.

Mostafa-Atallah2020 replied:
I've already added parameter checking to ensure gates are only merged if they have identical parameters. The implementation uses the [`_parameters_match()` method](https://github.com/Qiskit/qiskit/pull/15352/files#diff-5e3b3f7c8e1f0c8a1f7c8e1f0c8a1f7cR157-R178) which compares parameters within a configurable tolerance (default 1e-10), and the [`_group_compatible_gates()` method enforces this check](https://github.com/Qiskit/qiskit/pull/15352/files#diff-5e3b3f7c8e1f0c8a1f7c8e1f0c8a1f7cR262) before grouping gates together.

I've added comprehensive test coverage for parameter handling. The [`test_no_optimization_different_params()`](https://github.com/Qiskit/qiskit/pull/15352/files#diff-testfileR157-R176) test verifies that gates with different control patterns and different parameters (e.g., `RXGate(π/4).control(2, ctrl_state='11')` and `RXGate(π/3).control(2, ctrl_state='01')`) are correctly kept separate for Boolean algebra simplification.

Additionally, [`test_identical_patterns_different_params_merge()`](https://github.com/Qiskit/qiskit/pull/15352/files#diff-testfileR179-R199) tests the case where gates have identical control patterns but different rotation angles (e.g., `RXGate(π/4).control(2, ctrl_state='11')` and `RXGate(π/3).control(2, ctrl_state='11')`). These gates have the same control condition, so they can be left for other optimization passes (like gate combining) to merge into a single rotation with angle π/4 + π/3. This pass focuses on simplifying the control logic itself.

The parameter comparison logic uses [`np.isclose()`](https://github.com/Qiskit/qiskit/pull/15352/files#diff-5e3b3f7c8e1f0c8a1f7c8e1f0c8a1f7cR172) for numeric parameters and exact equality for symbolic parameters to ensure correct grouping.

Mostafa-Atallah2020 commented:
Hi @alexanderivrii, @Cryoris, and @debasmita2102 I totally removed the sympy dependency please rereview the PR

Mostafa-Atallah2020 commented:
@Cryoris linters issue fixed

alexanderivrii commented:
I had a brief look at the PR (so no detailed feedback yet), honestly it's significantly more code than I expected. At this point I am wondering if it would be better to have this as a Qiskit-Community project and not a part of Qiskit itself, however I do want to look at the code in more detail and discuss with others before making a particular suggestion.

I have a few high-level questions (in no particular order): 

1. What is the difference between the _complementary_ patterns and _subset_ patterns, it seems that the former is a special case of the latter, right? Also, why the latter optimization is called "subset", we are still considering terms with exactly the same control bits, i.e. not the following case when the control bits of one term are a subset of the other: ``(q0 ∧ q1 ∧ q2) ∨ (q0 ∧ ¬q2)``?

2. When reasoning about reductions, shouldn't we use ``⊕`` instead of ``∨``, i.e. ``(q0 ∧ q1) ⊕ (q0 ∧ ¬q1) = q0``instead of ``(q0 ∧ q1) ∨ (q0 ∧ ¬q1) = q0``? Here both are correct, but more generally if the terms are not disjoint, the original circuit can apply the controlled gate twice while the simplified circuit cannot. Do you know of any reductions when the terms are not necessarily disjoint?

3. Since this is application-driven (you mentioned quantum walks in the issue), what is the typical number of control qubits ? Realistically, do we expect to see about say at most 5 controls for each gate or hundreds? This has an effect of whether we can construct a truth table and you the already existing `EsopGenerator` code; @gadial, what do you think about this?

4. The XOR pairs simplification is interesting, I did not know about this before, is there a reference for this? How did you originally plan to use `sympy` for this? Also the optimization can probably be generalized, for instance the following reduction is also valid:
```
q_0: ─────o──────────■─────
          │          │     
q_1: ─────o──────────■─────
          │          │     
q_2: ─────o──────────■─────
     ┌────┴────┐┌────┴────┐
q_3: ┤ Rx(0.1) ├┤ Rx(0.1) ├
     └─────────┘└─────────┘
                                    
q_0: ──■────■───────────────■────■──
     ┌─┴─┐  │               │  ┌─┴─┐
q_1: ┤ X ├──┼───────o───────┼──┤ X ├
     └───┘┌─┴─┐     │     ┌─┴─┐└───┘
q_2: ─────┤ X ├─────o─────┤ X ├─────
          └───┘┌────┴────┐└───┘     
q_3: ──────────┤ Rx(0.1) ├──────────
               └─────────┘          
```

5. Have you thought about the complexity of your approach? From a brief look, it seemed that you are finding one reduction opportunity at a time while each time analyzing a quadratic number of pairs to consider, this will definitely become very slow if you have many long control patterns.

Mostafa-Atallah2020 commented:
Hi @alexanderivrii,

Thank you for the detailed review. Here are my responses:

**Complementary vs Subset Patterns:** You're right that complementary patterns (Hamming distance 1) are a special case of subset patterns. I can consolidate these into a single detection method if preferred.

**Boolean Logic & Double Application:** The patterns represent mutually exclusive control states—no quantum state can match multiple patterns simultaneously, so there's no risk of double gate application.

**Complexity:** The typical use case involves 2-6 control qubits with 2-10 patterns (from quantum walk algorithms). The $O(n^2 \times m)$ bitwise approach, where $n$ is the number of patterns and $m$ is the number of control qubits, handles this efficiently. For edge cases, the implementation includes early termination, iterative reduction, and complete partition detection; pathological inputs return early rather than hanging.

**XOR Optimization:** The CX trick is documented in:
- Atallah et al., "Graph Matching Trotterization for Continuous Time Quantum Walk Circuit Simulation" (IEEE QCE 2025)
- Gonzalez et al., "Efficient sparse state preparation via quantum walks" (npj Quantum Information 2025)

**SymPy Dependency:** Completely removed—the implementation now uses pure bitwise operations.

**Unit Test:** I will add a test based on your 3-controlled RXs example.

**Code Reduction:** The current implementation (~1165 lines) can be reduced by ~30-40% by:
- Merging `_build_single_control_gate` and `_build_multi_control_gate`
- Consolidating `_build_pairwise_optimized_gates` and `_build_iterative_pairwise_gates`
- Unifying `xor_standard` and `xor_with_x` with a flag
- Removing unused `_hamming_distance` and duplicate code
- Combining complementary and subset detection as you noted

alexanderivrii commented:
Thanks @Mostafa-Atallah2020 for the answers, and I will take a look the above paper and your code (but in a week or so, after Qiskit 2.3 release). One question that you did not answer: would you have been even able to use sympy to find patterns that differ at two or more positions?

And a few more random questions, just by thinking about possible optimizations that one can perform.

1. would you be able to simplify a circuit with the following 3 gates:
* `RZ(theta)` controlled by `x=0`
* `RZ(theta)` controlled by `x=1 & y=0`
* `RZ(theta)` controlled by `x=1 & y=1`

Intuitively, the controls still form a complete partition.


2. would you be able to simplify a circuit with the following 3 gates:
* `RZ(theta)` controlled by `x=0 & y=0`
* `RZ(theta)` controlled by `x=1 & y=0`
* `RZ(theta)` controlled by `x=1 & y=1`

Here the _complement_ of the involved patterns is a single pattern `x=0 & y=1`, so potentially you could reduce the 3 gates above to `RZ(theta)` controlled by `x=0 & y=1` + some additional X-gates.

Mostafa-Atallah2020 commented:
Hi @alexanderivrii,

Thank you for the suggestions. I've implemented both optimizations and added unit tests. Both the original SymPy-based version and the current bitwise implementation support all these optimizations.

**Example 1: Mixed Control Counts**

```
RZ(θ) controlled by x=0
RZ(θ) controlled by x=1 & y=0
RZ(θ) controlled by x=1 & y=1
```

This simplifies to unconditional `RZ(θ)` because:
```
x=0 OR (x=1 & y=0) OR (x=1 & y=1) = TRUE
```

Result: 3 gates → 1 gate. Unit test: `test_mixed_control_counts_to_unconditional`

**Example 2: Complement Optimization**

```
RZ(θ) controlled by x=0 & y=0  → pattern 00
RZ(θ) controlled by x=1 & y=0  → pattern 10
RZ(θ) controlled by x=1 & y=1  → pattern 11
```

Patterns `[00, 10, 11]` missing `01`. This optimizes to:
```
RZ(θ)              # unconditional
RZ(-θ) ctrl=01     # cancels the missing pattern
```

Result: 3 gates → 2 gates. Unit test: `test_complement_example_00_10_11`

Cryoris commented:
@Mostafa-Atallah2020 FYI you don't have to keep merging from main, the bot will automatically update on approval. We actually prefer _not_ updating manually since that triggers CI and consumes resources 🙂

alexanderivrii commented on test/python/transpiler/test_control_pattern_simplification.py:
```diff
@@ -0,0 +1,1147 @@
+                1.0,
+                places=10,
+                msg=f"{msg}: Fidelity mismatch for basis state |{format(i, f'0{num_qubits}b')}⟩",
+            )
```
You can also construct and compare unitary matrices representing the two circuits, in particular the verification code becomes significantly faster.

```suggestion
        self.assertEqual(Operator(qc1), Operator(qc2), msg=msg)
```

alexanderivrii commented on qiskit/transpiler/passes/optimization/control_pattern_simplification.py:
```diff
@@ -0,0 +1,443 @@
+            for g in group:
+                dag.remove_op_node(g.node)
+            for gate, qargs in repl:
+                dag.apply_operation_back(gate, [dag.qubits[i] for i in qargs])
```
Note that the code always adds new gates to the end of the DAG which is generally incorrect. You can see this using the following example (a modified version of one of your tests):
```python

    theta = np.pi / 4

    unsimplified_qc = QuantumCircuit(3)
    unsimplified_qc.append(RXGate(theta).control(2, ctrl_state=2, annotated=False), [0, 1, 2])
    unsimplified_qc.append(RXGate(theta).control(2, annotated=False), [0, 1, 2])
    unsimplified_qc.y(2)
    print(unsimplified_qc)

    # Run optimization
    pass_ = ControlPatternSimplification()
    optimized_qc = pass_(unsimplified_qc)

    print(optimized_qc)
    print(Operator(optimized_qc) == Operator(unsimplified_qc))
```

The simplest solution is to construct a new dag (rather than modifying the existing one), initially calling `new_dag = dag.copy_empty_like()` and then appending gates to this new dag instead.

alexanderivrii commented on qiskit/transpiler/passes/optimization/control_pattern_simplification.py:
```diff
@@ -0,0 +1,443 @@
+        if isinstance(cs, str):
+            return cs[::-1]
+        if isinstance(cs, int):
+            return format(cs, f"0{num_ctrl}b")[::-1]
```
The method `gate.ctrl_state` always returns an integer, you can indeed construct a ControlledGate by specifying `None`, string, or integer, but internally it's always stored as an integer.

alexanderivrii commented on qiskit/transpiler/passes/optimization/control_pattern_simplification.py:
```diff
@@ -0,0 +1,443 @@
+                        ctrl_state=self._extract_pattern(node.op, nc),
+                        params=tuple(node.op.params) if node.op.params else (),
+                    )
+                )
```
As mentioned before, you may want to extend the pass to handle `AnnotatedOperation`s as well.

alexanderivrii commented on qiskit/transpiler/passes/optimization/control_pattern_simplification.py:
```diff
@@ -0,0 +1,443 @@
+    control_qubits: List[int]
+    target_qubits: List[int]
+    ctrl_state: str
+    params: Tuple[float, ...]
```
Do you need to introduce this class, or would it be sufficient to just keep the node index in the DAG?

Mostafa-Atallah2020 replied:
you're right I will remove it

alexanderivrii reviewed (commented):
Thanks again for the contribution and sorry for the slow review.

A few general comments and questions.

In Qiskit 3.0 we want to make the change that `gate.control` by default produces an `AnnotatedOperation` rather than a `ControlledGate` (working with annotated operations provides many additional reduction opportunities especially for hierarchical circuits). As of now, however, `gate.control` still produces a `ControlledGate` but raises a deprecation warning.

I think that extending the code to handle `AnnotatedOperation`s  should not be too difficult, but I have not thought through all the details. So if this turns out to be more subtle, we can finalize the pass for `ControlledGate`s first, and extend it to handle `AnnotatedOperation`s at the end. In this case, to make testing happy, you need to add the argument `annotated=False` to every place where you call the `control` method (both in the transpiler pass itself and in the test file).

Did you use LLMs for any part of this work?

Have you considered working directly with controlled states represented as integers rather than as bitstrings? You would then be able to use bitwise arithmetic for things like finding the bits where two control states differ.

I have left a few additional comments and questions in place, and I will probably have more when I start looking at the code in more details.

alexanderivrii commented on qiskit/transpiler/passes/optimization/control_pattern_simplification.py:
```diff
@@ -0,0 +1,443 @@
+            if len(group) >= 2:
+                groups.append(group)
+            i = j if j > i + 1 else i + 1
+        return groups
```
Even though this is an internal function, having a clear docstring/code comments would enormously help: what are the inputs and outputs to this functions? what does the function do? how does it do it?  _This applies to many other functions in this file._ 

I roughly understand that the idea is to partition a sequence of consecutive controlled gates into subsequences, where each subsequence can be possibly reduced. But I couldn't follows the details, and in fact found another example where the output of the pass is incorrect:
```python
    unsimplified_qc = QuantumCircuit(3)
    unsimplified_qc.append(RXGate(0.1).control(2, ctrl_state=2), [0, 1, 2])
    unsimplified_qc.append(RXGate(0.1).control(2), [1, 0, 2])
    unsimplified_qc.append(RXGate(0.1).control(2), [0, 1, 2])
    print(unsimplified_qc)

    optimized_qc = ControlPatternSimplification()(unsimplified_qc)
    print(optimized_qc)

    print(Operator(optimized_qc) == Operator(unsimplified_qc))
```

There is an important question here. This is already the second example where the pass produces an incorrect output but is not captured by tests. How should we improve our testing framework? I am thinking that we need to add some kind of "random" tests, but not completely sure how they should look like.

Back to the function, for each gate `i` the code attempts to group it with gate `i+1`, `i+2`, etc., and breaks as soon as the new gate cannot be added to the group, is this correct? In this case, this can be implemented using a linear pass, without the nested while loop.

One more question: the following circuit is not optimized:
```
q_0: ─────o──────────■──────────■─────
          │          │          │     
q_1: ─────■──────────■──────────■─────
     ┌────┴────┐     │     ┌────┴────┐
q_2: ┤ Rx(0.1) ├─────┼─────┤ Rx(0.1) ├
     └─────────┘┌────┴────┐└─────────┘
q_3: ───────────┤ Rx(0.1) ├───────────
                └─────────┘           

```
even though the first and the third gates can be combines, and the second gate commutes with either of them. I believe it would be nice to support this.

Mostafa-Atallah2020 commented:
@alexanderivrii Thank you for the detailed review and no worries at all about the timing.

The pass now handles `AnnotatedOperation` with `ControlModifier` in `_get_controlled_info()`, so it is ready for the Qiskit 3.0 transition. All `.control()` calls in both the pass and tests use `annotated=False` to avoid the current deprecation warning. The `ctrl_state` is already used as an integer internally, and the pass uses bitwise AND, XOR, and bit counting for all pattern analysis including finding differing bits, common factors, and complement detection, with no bitstring conversions in the core logic. Regarding LLMs, I used them only for hints when I got stuck on specific issues, not for generating the core implementation or algorithm design. Tests use `self.assertEqual(Operator(qc1), Operator(qc2))` as you suggested, and DAG construction uses the `dag.copy_empty_like()` + `apply_operation_back()` pattern. I also simplified `_get_controlled_info` by removing unnecessary `None`/`str` checks for `ctrl_state` since it always returns `int` in both `ControlledGate` and `ControlModifier`.

1ucian0 committed:
squash

Mostafa-Atallah2020 committed:
Fixed: set comprehension issue

Mostafa-Atallah2020 committed:
Fixed: header issue
